File: FirstAkkaApplication/src/main/java/akka/first/app/mapreduce/scala/actors/MasterActor.scala
package akka.first.app.mapreduce.scala.actors

import akka.actor.actorRef2Scala
import akka.actor.Actor
import akka.actor.Props
import akka.first.app.mapreduce.scala.{MapData, ReduceData, Result}
import akka.routing.RoundRobinRouter

class MasterActor extends Actor {

	val mapActor = context.actorOf(Props[MapActor].withRouter( RoundRobinRouter(nrOfInstances = 5)), name = "map")
	val reduceActor = context.actorOf(Props[ReduceActor].withRouter( RoundRobinRouter(nrOfInstances = 5)), name = "reduce")
	val aggregateActor = context.actorOf(Props[AggregateActor], name = "aggregate")

	def receive: Receive = {
		case line: String =>
			mapActor ! line
		case mapData: MapData =>
			reduceActor ! mapData
		case reduceData: ReduceData =>
			aggregateActor ! reduceData
		case Result =>
			aggregateActor forward Result
	}
}

File: FirstAkkaApplication/src/main/java/akka/first/app/mapreduce/java/actors/MasterActor.java
package akka.first.app.mapreduce.java.actors;

import akka.actor.ActorRef;
import akka.actor.Props;
import akka.actor.UntypedActor;
import akka.first.app.mapreduce.java.message.MapData;
import akka.first.app.mapreduce.java.message.ReduceData;
import akka.first.app.mapreduce.java.message.Result;
import akka.first.app.mapreduce.java.message.*;
import akka.first.app.mapreduce.java.actors.ReduceActor;
import akka.routing.RoundRobinRouter;

public class MasterActor extends UntypedActor {

    private ActorRef mapActor = getContext().actorOf(new Props(MapActor.class). withRouter(new RoundRobinRouter(5)), "map" );
    private ActorRef reduceActor = getContext().actorOf(new Props(ReduceActor.class). withRouter(new RoundRobinRouter(5)), "reduce");
    private ActorRef aggregateActor = getContext().actorOf( new Props(AggregateActor.class), "aggregate");

    @Override
    public void onReceive(Object message) throws Exception {
        if (message instanceof String) {
            mapActor.tell(message,getSelf());
        } else if (message instanceof MapData) {
            reduceActor.tell(message,getSelf());
        } else if (message instanceof ReduceData) {
            aggregateActor.tell(message);
        } else if (message instanceof Result) {
            aggregateActor.forward(message, getContext());
        } else
            unhandled(message);
    }
}
